### PR TITLE
Feature: Introduce additional form params (#5399)

### LINKED ISSUE #5399: Feature request: add `additionalFormParams` to `authConfigs`
In [oauth2-authorize.js](https://github.com/swagger-api/swagger-ui/blob/1e184e8e218676278c83e60a45846c199ce3d15e/src/core/oauth2-authorize.js#L69) the `additionalQueryStringParams` property is used from the `authConfigs` object in order to add custom query string parameters to an authentication request.

Could we get a similar `additionalFormParams` property added that could then be passed to the `authActions` via method params or a property? That would allow [actions.js](https://github.com/swagger-api/swagger-ui/blob/937c8f6208f3adf713b10a349a82a1b129bd0ffd/src/core/plugins/auth/actions.js#L124) to add custom form properties as it's building the form data string.

**Reason:**
We're trying to use Swagger-UI (via [Swashbuckle.AspNetCore](https://github.com/domaindrivendev/Swashbuckle.AspNetCore)) to authenticate with Auth0. It has the right mix of params for the `implicit` grant type, but we're finding that `client_credentials` and `password` out-of-the-box body params passed aren't meeting Auth0's requirements. For example: when trying to use `client_credentials`, Auth0 requires `audience` in the body, but swagger-ui only passes `grant_type` and `scope`. Adding `audience` to the query string params doesn't cut it. This is where people are talking about getting the "Non-global clients are not allowed access to APIv1" error from Auth0.

If we had a similar way to add custom form/body params like we do query string params, it would make life easier.

Cross-post side note: ideally the Swashbuckle.AspNetCore library could then add another extension method similar to [`OAuthAdditionalQueryStringParams`](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/c1fb708e3fe3f2c62abf4a85b91786f38b66582e/src/Swashbuckle.AspNetCore.SwaggerUI/SwaggerUIOptionsExtensions.cs#L244)  that would feed those vales through in a similar way.

### PR DESCRIPTION
**Pull request by wdolek**
#### `authorizeRequest` change:

Changing "signature" of `authorizeRequest()` in `src/core/plugins/auth/actions.js`:

Instead of `body string`, we expect `form object` - reason is that this method already sets request headers, setting `Content-Type: application/x-www-form-urlencoded`. There's  no reason to split responsibility to multiple methods.

I checked whether `authorizeRequest` is used anywhere else, but it is used only from other methods within same file (despite it is exported).

This change has been done to simplify further work.

#### `additionalFormParams` property:

New property introduced in schema, `additionalFormParams` which can be used for adding custom parameters into authentication payload. This is required by #5399 - read issue for more details.

### Motivation and Context

Allows user to specify custom parameters in request payload. For example Auth0 requires `audience` in request payload, which is currently not possible to add.

This PR resolves #5399.

### How Has This Been Tested?

I created `swagger.json` containing (real URI replaced by example):

```
  ...
  "securityDefinitions": {
    "LeSec": {
      "flow": "application",
      "tokenUrl": "https://oauth.example.com/v2/token",
      "type": "oauth2",
      "description": "Authentication with Acme security",
      "additionalFormParams": {
        "audience": "https://api.example.com/"
      }
    }
  },
  ...
```

1. tested against `npm run dev`
2. tested against `dist` after `npm run build`

I was able to get valid access token to our service.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [x] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [ ] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [x] My changes require a change to the project documentation.
- [x] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ahuston-indigoag** [comment]
This is a useful change; there are many users of Auth0 who are effectively blocked from using the swagger-ui with clientCredentials. Is there an estimate of when this may be merged to swagger-api:master?

**Comment by ahuston-indigoag** [comment]
This will break openapi spec validators like https://github.com/p1c2u/openapi-spec-validator/blob/master/openapi_spec_validator/resources/schemas/v3.0.0/schema.json . Why not use x-additionalFormParams?

**Comment by wdolek** [comment]
@ahuston-indigoag Hi, will check that, I wasn't aware about schema validation - silly me.

**Comment by ahuston-indigoag** [comment]
> What would be the best approach here? Having valid schema even when it can't be valid by design or to introduce inconsistency?

@wdolek You are correct that this would introduce an inconsistency. Also, I suspect that swagger.io would like to keep to the current standard for their extensions, e.g. "additionalQueryParams", "additionalFormParams" etc.

Given that the current naming standard for the extensions is not likely to change, my advice would be to follow the Swagger.io naming standard. SmartBear would do well to maintain a branch of https://github.com/p1c2u/openapi-spec-validator/blob/master/openapi_spec_validator/resources/schemas/v3.0.0/schema.json that addresses the differences.

The other possibility would be for somebody to fork the swagger-ui branch to use "x-" but I think that would be less workable. Better to extend the 'standard' in openapi-spec-validator.

The longer term hope is that these new fields will make their way into the openapi spec for 3+ or 4+. SmartBear, if you are listening, is this the intent?

Thanks @wdolek for submitting this PR. Nice work!

File: docs/usage/oauth2.md [hunk 1]
```diff
@@ -23,6 +24,7 @@ ui.initOAuth({
     appName: "your-app-name",
     scopeSeparator: " ",
     additionalQueryStringParams: {test: "hello"},
+    additionalFormParams: {audience: "me"}
```
Comment (justinttl) [inline docs/usage/oauth2.md]: I think you are missing a trailing comma here

**Comment by justinttl** [comment]
Hi @wdolek, 

Does this also work with openapi version 3.0 `swagger.json`?

**Comment by webron** [comment]
Thanks for the PR. At the moment, there's going to be a delay in reviewing it, unfortunately.

Please note that `openapi-spec-validtor` is not an official tool, and the schema it uses can not be used alone to validate API definitions.

SmartBear does not control the spec. For feature request, please file tickets at https://github.com/OAI/OpenAPI-Specification.

Keep in mind there's a difference between introducing new keywords in the definition itself and adding more configuration options to Swagger UI. The latter is OK, for the most part, if there's a reasonable use case. The former should only be done using extensions, and support for such extensions should be added via plugins and not directly into the core code of the project.

File: docs/usage/oauth2.md (hunk 1 above)
Comment (wdolek) [reply docs/usage/oauth2.md]: Nice catch!

**Comment by wdolek** [comment]
> Does this also work with openapi version 3.0 `swagger.json`?

@justinttl Hi, to be honest, I'm not sure. My experience with Swagger is limited - how can I switch (and test) OpenAPI support?

**Comment by ThoSap** [comment]
Is this PR planned to be merged?

**Comment by wdolek** [comment]
@ThoSap Hi, I tried rebasing but apparently more things have changed - after all, it was sitting here for one year. I don't have time to investigate further at the moment, I would even suggest to start over - maybe there's better way how to achieve this today.

If anyone interested in this task, check ticket as well: #5399 

(For us, we perceive Swagger solely as documentation now - we have abandoned it as tool for making requests)

**Pull request closed without merging**